src/bind.py

```diff
@@ -169,8 +195,64 @@ def update_zonefiles_and_reload(
         # Reload charmed-bind config (only if already started)
         self.reload(force_start=False)
 
-        # Return the provider data with the entries' status
-        return self._create_dns_record_provider_data(relation_data)
+    def create_dns_record_provider_data(
+        self,
+        relation_data: list[tuple[DNSRecordRequirerData, DNSRecordProviderData]],
+    ) -> DNSRecordProviderData:
+        """Create dns record provider data from relation data.
+
+        Args:
+            relation_data: input relation data
+
+        Returns:
+            A DNSRecordProviderData object with requests' status
+        """
+        zones = self._dns_record_relations_data_to_zones(relation_data)
+        nonconflicting, conflicting = self._get_conflicts(zones)
+        statuses = []
+        for record_requirer_data, _ in relation_data:
+            for requirer_entry in record_requirer_data.dns_entries:
+                dns_entry = create_dns_entry_from_requirer_entry(requirer_entry)
+                if dns_entry in nonconflicting:
+                    statuses.append(
+                        DNSProviderData(uuid=requirer_entry.uuid, status=Status.APPROVED)
+                    )
+                    continue
+                if dns_entry in conflicting:
+                    statuses.append(
+                        DNSProviderData(uuid=requirer_entry.uuid, status=Status.CONFLICT)
+                    )
+                    continue
+                statuses.append(DNSProviderData(uuid=requirer_entry.uuid, status=Status.UNKNOWN))
+        return DNSRecordProviderData(dns_entries=statuses)
+
+    def has_a_zone_changed(
+        self,
+        relation_data: list[tuple[DNSRecordRequirerData, DNSRecordProviderData]],
+    ) -> bool:
+        """Check if a zone definition has changed.
+
+        Args:
+            relation_data: input relation data
+
+        Returns:
+            True if a zone has changed, False otherwise.
+        """
+        zones = self._dns_record_relations_data_to_zones(relation_data)
+        for zone in zones:
+            zonefile_content = pathlib.Path(
+                constants.DNS_CONFIG_DIR, f"db.{zone.domain}"
+            ).read_text(encoding="utf-8")
+            try:
+                metadata = self._get_zonefile_metadata(zonefile_content)
+            except (
+                exceptions.InvalidZoneFileMetadataError,
+                exceptions.EmptyZoneFileMetadataError,
+            ):
+                return True
+            if "HASH" in metadata and hash(zone) != int(metadata["HASH"]):
```

Comment:
Reading the documentation, the hash of a string changes between different invocations of Python  https://docs.python.org/3/reference/datamodel.html#object.__hash__

Could that be an issue?

Reply:
I have implemented the hash function myself [here](https://github.com/canonical/bind-operator/pull/35/files#diff-f4ca477841bd3ad5c0b292cbfff7c3ae403a65fb6c97deb15ad996f85dc87bd8R64)

Reply:
Oh, but I see that it could still be an issue...

Reply:
I changed my hash implementation to use hashlib instead. I tried to keep the same interface (outputting an int). Is that a good idea or should I use bytes instead ?

Reply:
I discovered that `__hash__()` needs to output an int. So be it